### ISSUE TITLE
Ball pipeline does not shift a ball to another layer

### ISSUE THREAD

**Issue by loveaabb** [Bithack/principia#69]
### Version
Principia 1.5.2 Beta
### Platform
Windows 11
### Bug description
the ball pipeline fails to shift a ball to another layer, as if it is blocked by an object in the target layer
### Screenshots
![pipelinebug](https://user-images.githubusercontent.com/30974194/208138594-b8da3d17-170c-4dab-90e1-cb545d481050.png)

**Comment by loveaabb** [comment]
found a workaround
https://principia-web.se/level/488

**Comment by ManEnfu** [comment]
I encountered the same issue and it seems like the cause is incorrect floating point calculation due to optimization from `-ffast-math` compiler flag. I found a fix by adding `-fno-associative-math` flag.

**Comment by ManEnfu** [comment]
Suction cups also has similar issue, which the fix above solves as well.

**Comment by rollerozxa** [comment]
Is there even a bug here? I cannot reproduce this, suction cups work fine too.

**Comment by ManEnfu** [comment]
I think it can happen if your compiler have some exotic options enabled. In the case of the the ball pipeline, it is caused by `tmath_sqrt()` function in `src/tms/math/math.c`. Specifically, `tmath_sqrt(0.0)` returns `NaN` instead of `0.0` which messes up the pipeline internal logic.

**Comment by rollerozxa** [comment]
Okay I managed to reproduce it, it had already been fixed on Linux by reducing the optimisation level entirely so I wasn't able to reproduce it there but can in fact reproduce it on Windows. The best idea might be to fix it it with a check in `tmath_sqrt()` so as to keep the optimisation level on Windows and also increase it back on Linux. Though, I'm growing increasingly concerned about the inconsistency in optimisations between the three platforms not only causing issues like this but also leading to slight differences in physics behaviour that make some levels completely broken on particular platforms. (Though this has already happened, the physics never really stayed consistent in the past either)
